Improved header parsing for `ServerRequest::fromGlobals` (fixes guzzle/psr7#160)

kael-shipman opened the pull request.

kael-shipman commented:
Woops, I guess I should have looked more closely at your outstanding PRs. I see this is a duplicate of #147 . I'll leave it open so you can choose which one you prefer.

kael-shipman commented:
@Nyholm and/or @Tobion, please offer comments as to what needs to change to get this (or another comparable fix) merged. (Thanks)

sagikazarmark commented on src/ServerRequest.php:
```diff
@@ -166,7 +166,26 @@ private static function normalizeNestedFileSpec(array $files = [])
-        $headers = function_exists('getallheaders') ? getallheaders() : [];
+
+        $headers = [];
+        if (function_exists('getallheaders')) $headers = getallheaders();
```
Can you please follow PSR-1,2?

sagikazarmark commented on src/ServerRequest.php:
```diff
@@ -166,7 +166,26 @@ private static function normalizeNestedFileSpec(array $files = [])
+                if (substr($k, 0, 5) == 'HTTP_') {
+                    $k = explode('_', substr($k, 5));
+                    $header = [];
+                    foreach($k as $word) $header[] = ucfirst(strtolower($word));
```
Same here, coding style

sagikazarmark replied:
AFAIK headers are technically case-insensitive, so I'm not sure this is necessary. Also, in case of headers like `Content-MD5` or `X-Request-ID` it won't have the desired effect anyway.

But I would love to hear other opinions.

sagikazarmark commented:
Had a quick view, but I would love someone else to take a look as well.

Nyholm commented:
Thank you. Could you prove that this code works with a unit test?

sagikazarmark commented on src/ServerRequest.php:
```diff
@@ -166,7 +166,29 @@ private static function normalizeNestedFileSpec(array $files = [])
+                    $k = explode('_', substr($k, 5));
+                    $header = [];
+                    foreach($k as $word) {
+                        $header[] = ucfirst(strtolower($word));
```
This one lost, so reposting:

AFAIK headers are technically case-insensitive, so I'm not sure this is necessary. Also, in case of headers like `Content-MD5` or `X-Request-ID` it won't have the desired effect anyway.

But I would love to hear other opinions.

pavarnos commented:
have you seen https://github.com/ralouphie/getallheaders ? It does a few extra things like Authorization headers.

kael-shipman commented:
@pavarnos Thanks for the tip! I need to get better about noticing when other people have solved easy problems in a publicly-consumable way....

@Nyholm Sorry it took me a month and a half to do it, but I've added some extra lines to the test for `fromGlobals` to cover this. I've also reverted my changes back to the method's original state and instead added `ralouphie/getallheaders` polyfill as a dependency. Furthermore, I verified that the tests I added did indeed fail without the dependency, then succeeded with it.

Let me know if you need anything else.

GrahamCampbell commented on src/ServerRequest.php:
```diff
@@ -166,29 +166,7 @@ private static function normalizeNestedFileSpec(array $files = [])
-            }
-        }
-
+        $headers = function_exists('getallheaders') ? getallheaders() : [];
```
No need for that `function_exists` call anymore.

kael-shipman commented:
Silly me ;)

zuoRambo commented:
vote for this PR

zuoRambo commented:
When to release?
